File: repo/frontend/src/types/attendance.ts
export type AnomalyType = 'absent' | 'late' | 'unauthorized_absence' | 'other'

export type ExceptionStatus =
  | 'pending_proof'
  | 'pending_initial_review'
  | 'pending_final_review'
  | 'approved'
  | 'rejected'

export type ReviewStage = 'initial' | 'final'

export interface AttendanceAnomaly {
  id: string
  candidate_id: string
  anomaly_type: AnomalyType
  session_date: string
  description: string | null
  flagged_by: string
  flagged_at: string
  created_at: string
}

export interface AnomalyCreate {
  candidate_id: string
  anomaly_type: AnomalyType
  session_date: string
  description?: string | null
}

export interface ExceptionProof {
  id: string
  exception_id: string
  document_version_id: string
  uploaded_by: string
  description: string | null
  created_at: string
}

export interface ExceptionApproval {
  id: string
  step_id: string
  exception_id: string
  approved_by: string
  outcome: 'approved' | 'rejected' | 'escalated'
  signature_hash: string
  created_at: string
}

export interface ExceptionReviewStep {
  id: string
  exception_id: string
  step_order: number
  stage: ReviewStage
  reviewer_id: string
  reviewer_role: string
  decision: 'approve' | 'reject' | 'escalate'
  notes: string | null
  is_escalated: boolean
  reviewed_at: string
}

export interface AttendanceException {
  id: string
  anomaly_id: string | null
  candidate_id: string
  status: ExceptionStatus
  current_stage: ReviewStage
  candidate_statement: string
  proofs: ExceptionProof[]
  review_steps: ExceptionReviewStep[]
  created_at: string
  updated_at: string
}

export interface AttendanceExceptionCreate {
  candidate_statement: string
}

export interface ReviewDecisionCreate {
  decision: 'approve' | 'reject' | 'escalate'
  notes?: string | null
}

export interface ProofUploadResponse {
  proof_id: string
  document_version_id: string
  sha256_hash: string
}

File: repo/backend/src/schemas/attendance.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from ..domain.enums import AnomalyType, ExceptionStatus, ReviewDecision, ReviewStage


class AttendanceExceptionCreate(BaseModel):
    anomaly_id: uuid.UUID | None = None
    candidate_statement: str | None = Field(None, max_length=5000)


class AnomalyCreate(BaseModel):
    candidate_id: uuid.UUID
    anomaly_type: AnomalyType
    session_date: datetime
    session_identifier: str | None = Field(None, max_length=200)
    description: str | None = None


class AnomalyRead(BaseModel):
    id: uuid.UUID
    candidate_id: uuid.UUID
    anomaly_type: AnomalyType
    session_date: datetime
    session_identifier: str | None
    flagged_by: uuid.UUID
    flagged_at: datetime
    description: str | None


class ProofUploadResponse(BaseModel):
    proof_id: uuid.UUID
    exception_id: uuid.UUID
    document_version_id: uuid.UUID
    uploaded_at: datetime


class ReviewDecisionCreate(BaseModel):
    decision: ReviewDecision
    notes: str | None = Field(None, max_length=3000)


class ExceptionReviewStepRead(BaseModel):
    id: uuid.UUID
    exception_id: uuid.UUID
    step_order: int
    stage: ReviewStage
    reviewer_id: uuid.UUID
    reviewer_role: str
    decision: ReviewDecision
    notes: str | None
    decided_at: datetime
    is_escalated: bool


class ExceptionApprovalRead(BaseModel):
    id: uuid.UUID
    step_id: uuid.UUID
    exception_id: uuid.UUID
    approved_by: uuid.UUID
    approved_at: datetime
    outcome: str
    signature_hash: str | None


class ExceptionRead(BaseModel):
    id: uuid.UUID
    anomaly_id: uuid.UUID | None
    candidate_id: uuid.UUID
    status: ExceptionStatus
    current_stage: ReviewStage
    candidate_statement: str | None
    submitted_at: datetime | None
    created_at: datetime
    updated_at: datetime
    review_steps: list[ExceptionReviewStepRead] = []
